src/dbally/views/freeform/text2sql/_autodiscovery.py

```diff
@@ -0,0 +1,328 @@
+                "Determine whether to use SEMANTIC or TRIGRAM search based on the given column.\n"
+                "Use SEMANTIC when values are categorical or synonym may be used (for example department names).\n"
+                "Use TRIGRAM when values are sensitive to typos and synonyms does not make sense "
+                "(for example person or company names).\n"
+                "Use NONE when a search of the values does not make sense and explicit values should be utilized.\n"
+                "Return only one of the following options: SEMANTIC, TRIGRAM, or NONE."
+            ),
+        },
+        {
+            "role": "user",
+            "content": "TABLE SUMMARY: {table_summary}\n" "COLUMN_NAME: {column_name}\n" "EXAMPLE_VALUES: {values}",
+        },
+    )
+)
+
+
+class _Text2SQLAutoDiscovery:
+    """
+    Class for auto-discovery of the database for text2sql freeform view.
+    """
+
+    def __init__(
+        self,
+        engine: Engine,
+        description_extraction: _DescriptionExtractionStrategy,
+        whitelist: Optional[List[str]] = None,
+        llm_client: Optional[LLMClient] = None,
+        blacklist: Optional[List[str]] = None,
+        similarity_enabled: bool = False,
+    ) -> None:
+        self._llm_client = llm_client
+        self._engine = engine
+        self._whitelist = whitelist
+        self._blacklist = blacklist
+        self._description_extraction = description_extraction
+        self._similarity_enabled = similarity_enabled
+
+    async def discover(self) -> Text2SQLConfig:
+        """
+        Discover tables in the database and return the configuration object.
+
+        Returns:
+            Text2SQLConfig: The configuration object for the text2sql freeform view.
+
+        Raises:
+            ValueError: If the description extraction strategy is invalid.
+        """
+
+        connection = self._engine.connect()
+        tables = {}
+
+        for table_name, table in self._iterate_tables():
+            if self._whitelist is not None and table_name not in self._whitelist:
+                continue
+
+            if self._blacklist is not None and table_name in self._blacklist:
+                continue
+
+            ddl = self._get_table_ddl(table)
+
+            if isinstance(self._description_extraction, _DBCommentsDescriptionExtraction):
+                description = table.comment
+            elif isinstance(self._description_extraction, _LLMSummaryDescriptionExtraction):
+                example_rows = self._get_example_rows(connection, table, self._description_extraction.example_rows_cnt)
+                description = await self._generate_llm_summary(ddl, example_rows)
+            else:
+                raise ValueError(f"Invalid description extraction strategy: {self._description_extraction}")
+
+            if self._similarity_enabled:
+                similarity = await self._suggest_similarity_indexes(connection, description or "", table)
```

Comment:
I would probably skip the call to the LLM altogether in cases when there are more than n unique values in a given column

Reply:
Why? TRIGRAM method can be still relevant even for columns with high cardinality. But yeah, some logic may be introduce to reduce llm calls here, I think we can have a closer look on that later when we will be implementic full integration between text2sql view and similarity indexes.